### PR TITLE
add PLaMo model

### PR DESCRIPTION
**Pull request by okdshin**
This PR adds support of [PLaMo-13B](https://huggingface.co/pfnet/plamo-13b) model.

> PLaMo-13B is a LLaMA-based 13B model pre-trained on English and Japanese open datasets, developed by Preferred Networks, Inc. PLaMo-13B is released under Apache v2.0 license.

```
python convert-plamo-hf-to-gguf.py ./plamo-13b 1
./main -m ./plamo-13b/ggml-model-f16.gguf -p "これからの人工知能技術は" --temp 1.0 --top-p 0.95 --top-k 50 -n 32
```

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -4899,6 +4993,388 @@ static struct ggml_cgraph * llm_build_persimmon(
...
+                        ggml_element_size(kv_self.k)*n_embd_gqa,
+                        ggml_element_size(kv_self.k)*n_embd_head,
+                        ggml_element_size(kv_self.k)*n_embd_gqa*n_ctx*il);
+            offload_func_kq(K);
+            ggml_set_name(K, "K");
+
+            // K * Q
+            //struct ggml_tensor * KQ = ggml_mul_mat(ctx0, K, Q);
+            // we should avoid to repeat K but current ggml_mul_mat generates wrong values for grouped query att 
+            struct ggml_tensor * K_repeated = ggml_new_tensor_3d(ctx0, GGML_TYPE_F32, K->ne[0], K->ne[1], Q->ne[2]);
```
Comment (okdshin) [inline llama.cpp:5249]: plamo-13b uses grouped query attention. I tried `ggml_mul_mat(ctx0, K, Q);` but it generates wrong values so I repeated K.

Comment (ggerganov) [reply llama.cpp:5253]: Broadcast in `ggml_mul_mat` is supported, so this repeat should not be necessary. LLaMA v2 70B definitely works.

I can't see an immediate problem in the implementation - will need to get the model and do some tests

Comment (okdshin) [reply llama.cpp:6012]: Thank you for your reply. I'll add the data that shows the difference in values and upload gguf files later.

**Comment by ggerganov** [comment]
As far as I can tell, this is exactly the same architecture as LLaMA v2. We should try to fit this into the existing arch implementation instead of creating a new arch. Whatever issues with the GQA might exist (i.e. convert script, broadcast, etc.), they should be identified and resolved as part of the `LLM_ARCH_LLAMA`

**Comment by okdshin** [comment]
I uploaded plamo-13b gguf file. https://huggingface.co/sokada4/plamo-13b-gguf/tree/main

## w/ K repeating

```
> ./bin/main -m plamo-13b-gguf/ggml-model-Q4_1.gguf -p "Hello my name is" --temp 0 -n 32
...
 Hello my name is sara and I'm a student at the University of Toronto. I have been playing piano for about five years now, but I started when I
...
```

## w/o K repeating

```
git diff                                                                               
diff --git a/llama.cpp b/llama.cpp
index 9868c81..0ba534c 100644
--- a/llama.cpp
+++ b/llama.cpp
@@ -6007,12 +6007,14 @@ static struct ggml_cgraph * llm_build_plamo(
             ggml_set_name(K, "K");
 
             // K * Q
-            //struct ggml_tensor * KQ = ggml_mul_mat(ctx0, K, Q);
+            struct ggml_tensor * KQ = ggml_mul_mat(ctx0, K, Q);
             // we should avoid to repeat K but current ggml_mul_mat generates wrong values for grouped query att 
+            /*
             struct ggml_tensor * K_repeated = ggml_new_tensor_3d(ctx0, GGML_TYPE_F32, K->ne[0], K->ne[1], Q->ne[2]);
             offload_func_kq(K_repeated);
             ggml_set_name(K_repeated, "K_repeated");
             struct ggml_tensor * KQ = ggml_mul_mat(ctx0, ggml_repeat(ctx0, K, K_repeated), Q);
+            */
             offload_func_kq(KQ);
             ggml_set_name(KQ, "KQ");
```

```
> ./bin/main -m plamo-13b-gguf/ggml-model-Q4_1.gguf -p "Hello my name is" --temp 0 -n 32
...
 Hello my name is not a .вЖ, and then





</p>
...
```

**Comment by okdshin** [comment]
@ggerganov Excuse me. Please tell me what I have to do to merge this PR.

**Comment by cebtenzzre** [comment]
> Excuse me. Please tell me what I have to do to merge this PR.

I can't promise anything, but it would help if you resolved the merge conflicts.

**Comment by ggerganov** [comment]
@okdshin The model graph code in `llama.cpp` has been recently refactored (#3837) in order to reduce code duplication and make things more consistent. To merge this PR, it will have to be reworked to follow the new graph building pattern that is used for all other models.

**Comment by okdshin** [comment]
@cebtenzzre @ggerganov Thank you for your comments. I will modify this PR to follow the new graph building pattern.

**Comment by okdshin** [comment]
@ggerganov @cebtenzzre I reworked to follow  the new graph building pattern. Could you check it?

**Comment by ggerganov** [comment]
I was curious why the broadcast does not work here, so I tested the model and the following patch makes it work on the CPU, using the original `llm_build_kqv`:

```diff
diff --git a/ggml.c b/ggml.c
index ad546a73..026c2205 100644
--- a/ggml.c
+++ b/ggml.c
@@ -9650,7 +9650,7 @@ static void ggml_compute_forward_mul_mat(
             for (int64_t i12 = 0; i12 < ne12; i12++) {
                 // broadcast src0 into src1 across 2nd,3rd dimension
                 const int64_t i03 = i13/r3;
-                const int64_t i02 = i12/r2;
+                const int64_t i02 = i12%5;
 
                 const void  * x = (char *)            src0->data + i02*nb02 + i03*nb03;
                 const float * y = (float *) ((char *) src1->data + i12*nb12 + i13*nb13);
@@ -9761,7 +9761,7 @@ static void ggml_compute_forward_mul_mat(
 
                 // broadcast src0 into src1
                 const int64_t i03 = i13/r3;
-                const int64_t i02 = i12/r2;
+                const int64_t i02 = i12%5;
 
                 const int64_t i1 = i11;
                 const int64_t i2 = i12;
```

So it means we need during the conversion to shuffle the `attn_q.weight` tensors to match the expected order in `llama.cpp`. This will avoid having to implement the custom `plamo_llm_build_kqv` function.

**Comment by ggerganov** [comment]
I pushed the WIP version here: https://github.com/ggerganov/llama.cpp/tree/gg/plamo-test

This one works without explicit `ggml_repeat`, but obviously it's not OK because it includes the patch in `ggml.c` that I posted above. The goal is to update the convert script so that it shuffles the necessary tensors in such a way that we don't have to modify the broadcast logic in `ggml.c`. Hope this helps

**Comment by okdshin** [comment]
Thank you very much! I will modify the conversion script and try to avoid to use the customized `llm_build_kqv`.

**Comment by okdshin** [comment]
I modified to shuffle `attn_q.weight` and `attn_output.weight` for the broadcasting and removed `plamo_llm_build_kqv`. Could you review again?

**Review by ggerganov — APPROVED** [review]
Good job! Thank you for following up with all the necessary changes!

**Pull request merged**